File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/cards/CardSpoiler.java
package com.sup.dev.android.views.cards;

import android.graphics.PorterDuff;
import android.support.annotation.StringRes;
import android.text.Html;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.sup.dev.android.R;
import com.sup.dev.android.tools.ToolsResources;
import com.sup.dev.android.views.adapters.CardAdapter;

import java.util.ArrayList;

public class CardSpoiler extends Card {

    private final ArrayList<Card> cards = new ArrayList<>();
    private String title;
    private String text;
    private String rightText;
    private int titleColor = 0;
    private int rightTextColor = 0;
    private int textColor = 0;
    private boolean originalSeted = false;
    private boolean dividerVisible = true;
    private int titleColorOriginal;
    private int rightTextColorOriginal;
    private int textColorOriginal;
    private int iconColor = 0;

    boolean expanded;
    boolean enabled = true;

    @Override
    public int getLayout() {
        return R.layout.card_spoiler;
    }

    @Override
    public void bindView(View view) {
        ImageView vIcon = view.findViewById(R.id.icon);
        TextView vTitle = view.findViewById(R.id.title);
        TextView vText = view.findViewById(R.id.text);
        TextView vRightText = view.findViewById(R.id.right_text);
        View vTouch = view.findViewById(R.id.touch);
        View vDivider = view.findViewById(R.id.divider);

        if (!originalSeted) {
            originalSeted = true;
            titleColorOriginal = vTitle.getCurrentTextColor();
            textColorOriginal = vText.getCurrentTextColor();
            rightTextColorOriginal = vRightText.getCurrentTextColor();
        }

        vText.setText(text == null ? null : Html.fromHtml(text));
        vRightText.setText(rightText == null ? null : Html.fromHtml(rightText));
        vTitle.setText(title == null ? null : Html.fromHtml(title));

        vText.setVisibility(text == null ? View.GONE : View.VISIBLE);
        vRightText.setVisibility(rightText == null ? View.GONE : View.VISIBLE);
        vTitle.setVisibility(title == null ? View.GONE : View.VISIBLE);

        vText.setEnabled(enabled);
        vRightText.setEnabled(enabled);
        vTitle.setEnabled(enabled);

        vDivider.setVisibility(dividerVisible?View.VISIBLE:View.GONE);
        vText.setTextColor(textColor != 0 ? textColor : textColorOriginal);
        vRightText.setTextColor(rightTextColor != 0 ? rightTextColor : rightTextColorOriginal);
        vTitle.setTextColor(titleColor != 0 ? titleColor : titleColorOriginal);

        vIcon.setImageResource(expanded ? R.drawable.ic_keyboard_arrow_up_black_24dp : R.drawable.ic_keyboard_arrow_down_black_24dp);
        vIcon.setAlpha(enabled ? 255 : 106);
        if (iconColor != 0) vIcon.setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP);
        vTouch.setOnClickListener(!enabled ? null : v -> setExpanded(!expanded));
        vTouch.setClickable(enabled);
    }

    //
    //  Setters
    //

    public CardSpoiler add(Card card) {
        cards.add(card);
        setExpanded(expanded);
        return this;
    }

    public CardSpoiler setTitle(@StringRes int title) {
        return setTitle(ToolsResources.getString(title));
    }

    public CardSpoiler setTitle(String title) {
        this.title = title;
        update();
        return this;
    }

    @Override
    public void setAdapter(CardAdapter adapter) {
        super.setAdapter(adapter);
        setExpanded(expanded);
    }

    public CardSpoiler setExpanded(boolean expanded) {
        this.expanded = expanded;
        update();

        if(adapter != null) {
            if (expanded) {

                int myIndex = adapter.indexOf(this);
                for (Card c : cards)
                    if (myIndex != -1)
                        if (!adapter.contains(c)) adapter.add(++myIndex, c);

            } else
                for (Card c : cards) adapter.remove(c);
        }

        return this;
    }

    public CardSpoiler setEnabled(boolean enabled) {
        this.enabled = enabled;
        update();
        return this;
    }

    public CardSpoiler setText(String text) {
        this.text = text;
        update();
        return this;
    }

    public CardSpoiler setRightText(String rightText) {
        this.rightText = rightText;
        update();
        return this;
    }

    public CardSpoiler setIconColor(int iconColor) {
        this.iconColor = iconColor;
        update();
        return this;
    }

    public CardSpoiler setTextColor(int textColor) {
        this.textColor = textColor;
        update();
        return this;
    }

    public CardSpoiler setTitleColor(int titleColor) {
        this.titleColor = titleColor;
        update();
        return this;
    }

    public CardSpoiler setRightTextColor(int rightTextColor) {
        this.rightTextColor = rightTextColor;
        update();
        return this;
    }

    public CardSpoiler setDividerVisible(boolean dividerVisible) {
        this.dividerVisible = dividerVisible;
        update();
        return this;
    }

    //
    //  Getters
    //

    public ArrayList<Card> getCards() {
        return cards;
    }
}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/views/cards/CardSpoiler.kt
package com.sup.dev.android.views.cards

import android.graphics.PorterDuff
import android.support.annotation.StringRes
import android.text.Html
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.sup.dev.android.R
import com.sup.dev.android.tools.ToolsResources
import com.sup.dev.android.views.support.adapters.CardAdapter
import java.util.ArrayList


open class CardSpoiler : Card() {

    //
    //  Getters
    //

    val cards = ArrayList<Card>()
    private var title: String? = null
    private var text: String? = null
    private var rightText: String? = null
    private var titleColor = 0
    private var rightTextColor = 0
    private var textColor = 0
    private var originalSeted = false
    private var dividerVisible = true
    private var titleColorOriginal: Int = 0
    private var rightTextColorOriginal: Int = 0
    private var textColorOriginal: Int = 0
    private var iconColor = 0

    internal var expanded: Boolean = false
    internal var enabled = true

    override fun getLayout(): Int {
        return R.layout.card_spoiler
    }

    override fun bindView(view: View) {
        val vIcon = view.findViewById<ImageView>(R.id.vIcon)
        val vTitle = view.findViewById<TextView>(R.id.vTitle)
        val vText = view.findViewById<TextView>(R.id.vText)
        val vRightText = view.findViewById<TextView>(R.id.vRightText)
        val vTouch = view.findViewById<View>(R.id.vTouch)
        val vDivider = view.findViewById<View>(R.id.vDivider)

        if (!originalSeted) {
            originalSeted = true
            titleColorOriginal = vTitle.currentTextColor
            textColorOriginal = vText.currentTextColor
            rightTextColorOriginal = vRightText.currentTextColor
        }

        vText.text = if (text == null) null else Html.fromHtml(text)
        vRightText.text = if (rightText == null) null else Html.fromHtml(rightText)
        vTitle.text = if (title == null) null else Html.fromHtml(title)

        vText.visibility = if (text == null) View.GONE else View.VISIBLE
        vRightText.visibility = if (rightText == null) View.GONE else View.VISIBLE
        vTitle.visibility = if (title == null) View.GONE else View.VISIBLE

        vText.isEnabled = enabled
        vRightText.isEnabled = enabled
        vTitle.isEnabled = enabled

        vDivider.visibility = if (dividerVisible) View.VISIBLE else View.GONE
        vText.setTextColor(if (textColor != 0) textColor else textColorOriginal)
        vRightText.setTextColor(if (rightTextColor != 0) rightTextColor else rightTextColorOriginal)
        vTitle.setTextColor(if (titleColor != 0) titleColor else titleColorOriginal)

        vIcon.setImageResource(if (expanded) ToolsResources.getDrawableId("ic_keyboard_arrow_up") else ToolsResources.getDrawableId("ic_keyboard_arrow_down"))
        vIcon.setAlpha(if (enabled) 255 else 106)
        if (iconColor != 0) vIcon.setColorFilter(iconColor, PorterDuff.Mode.SRC_ATOP)
        if (enabled) vTouch.setOnClickListener{ v -> setExpanded(!expanded) }
        else vTouch.setOnClickListener(null)
        vTouch.isClickable = enabled
    }

    //
    //  Setters
    //

    fun add(card: Card): CardSpoiler {
        cards.add(card)
        setExpanded(expanded)
        return this
    }

    open fun setTitle(@StringRes title: Int): CardSpoiler {
        return setTitle(ToolsResources.getString(title))
    }

    fun setTitle(title: String?): CardSpoiler {
        this.title = title
        update()
        return this
    }

    override fun setCardAdapter(adapter: CardAdapter?) {
        super.adapter = adapter
        setExpanded(expanded)
    }

    fun setExpanded(expanded: Boolean): CardSpoiler {
        this.expanded = expanded
        update()

        if (adapter != null) {
            if (expanded) {

                var myIndex = adapter!!.indexOf(this)
                for (c in cards)
                    if (myIndex != -1)
                        if (!adapter!!.contains(c)) adapter!!.add(++myIndex, c)

            } else
                for (c in cards) adapter!!.remove(c)
        }

        return this
    }

    fun setEnabled(enabled: Boolean): CardSpoiler {
        this.enabled = enabled
        update()
        return this
    }

    fun setText(text: String): CardSpoiler {
        this.text = text
        update()
        return this
    }

    fun setRightText(rightText: String): CardSpoiler {
        this.rightText = rightText
        update()
        return this
    }

    fun setIconColor(iconColor: Int): CardSpoiler {
        this.iconColor = iconColor
        update()
        return this
    }

    fun setTextColor(textColor: Int): CardSpoiler {
        this.textColor = textColor
        update()
        return this
    }

    fun setTitleColor(titleColor: Int): CardSpoiler {
        this.titleColor = titleColor
        update()
        return this
    }

    fun setRightTextColor(rightTextColor: Int): CardSpoiler {
        this.rightTextColor = rightTextColor
        update()
        return this
    }

    fun setDividerVisible(dividerVisible: Boolean): CardSpoiler {
        this.dividerVisible = dividerVisible
        update()
        return this
    }
}
